File: MonadsForRealPeople.FSharp/EitherChoice.fs
module EitherChoice

open FsUnit
open NUnit.Framework
open ExtCore.Control

// setup
let mightFail v =
  if v > 3 then Choice.result ((v - 3).ToString()) else Choice.failwith "too small"

let mightAlsoFail v =
  match v with
  | "1" -> Choice.result [1]
  | "2" -> Choice.result [1 ; 2]
  | "3" -> Choice.result [1 ; 2 ; 3]
  | _ -> Choice.failwith "too big"

let couldFailToo v =
  match v with
  | [1 ; 2] -> Choice.result "OneTwoPunch"
  | _ -> Choice.failwith "Knockout"

let finalComputationOn x y z =
  sprintf "%s %A %s" x y z

// either monad

let eitherEither() : Choice<string,string> =
  choice {
    let! x = mightFail 5
    let! y = mightAlsoFail x
    let! z = couldFailToo y
    return finalComputationOn x y z
  }

// test
[<TestFixture>]
type ``test either``() =
  [<Test>]
  member x.``either works``() =
    eitherEither() |> Choice.get |> should equal "2 [1; 2] OneTwoPunch"

File: MonadsForRealPeople.CSharp/Longhand/EitherChoice.cs
using Monad;
using MonadsForRealPeople.CSharp;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using NUnit.Framework;
using MonadsForRealPeople.CSharp.Types;

namespace MonadsForRealPeople.CSharp.Longhand
{
    public class ChoiceExample
    {
        public static Choice<Exception,string> MightFail(int value)
        {
            return (value > 3) 
                ? Choice<Exception,string>.Success((value - 3).ToString())
                : Choice<Exception,string>.Fail(new Exception("too small"));
        }

        public static Choice<Exception,int[]> MightAlsoFail(string value)
        {
            switch (value)
            {
                case "1": return Choice<Exception,int[]>.Success(new[] { 1 });
                case "2": return Choice<Exception,int[]>.Success(new[] { 1, 2 });
                case "3": return Choice<Exception,int[]>.Success(new[] { 1, 2, 3 });
                default: return Choice<Exception,int[]>.Fail(new Exception("too big"));
            }
        }

        public static Choice<Exception,string> CouldFailToo(int[] input)
        {
            if (input.SequenceEqual(new [] { 1, 2}))
            {
                return Choice<Exception, string>.Success("OneTwoPunch");
            }
            else
            {
                return Choice<Exception, string>.Fail(new Exception("Knockout"));
            }
        }

        public static string FinalComputation(string x, int[] y, string z)
        {
            return string.Format("{0} [{1}] {2}", x, string.Join(",", y), z);
        }

        public static string DoStuffThatMightFail()
        {
            var x = MightFail(5);
            if (!x.IsSuccess) { return x.FailValue.Message; }
            var y = MightAlsoFail(x.SuccessValue);
            if (!y.IsSuccess) { return y.FailValue.Message; }
            var z = CouldFailToo(y.SuccessValue);
            if (!z.IsSuccess) { return z.FailValue.Message; }
            return FinalComputation(x.SuccessValue, y.SuccessValue, z.SuccessValue);
        }
    }

    [TestFixture]
    public class TestChoice
    {
        [Test]
        public void EitherWorks()
        {
            Assert.AreEqual("2 [1,2] OneTwoPunch", ChoiceExample.DoStuffThatMightFail());
        }
    }
}
